roboteq_diff_driver/src/driver.cpp

```diff
@@ -218,329 +24,132 @@ MainNode::MainNode() :
 }
 
-
-//
-// cmd_vel subscriber
-//
-
-void MainNode::cmdvel_callback( const geometry_msgs::Twist& twist_msg)
+void MainNode::rpm_callback(const roboteq_diff_msgs::WheelRates& rpm_msg)
 {
-
-  // wheel speed (m/s)
-  float right_speed = twist_msg.linear.x + track_width * twist_msg.angular.z / 2.0;
-  float left_speed = twist_msg.linear.x - track_width * twist_msg.angular.z / 2.0;
-#ifdef _CMDVEL_DEBUG
-ROS_DEBUG_STREAM("cmdvel speed right: " << right_speed << " left: " << left_speed);
-#endif
-
   std::stringstream right_cmd;
   std::stringstream left_cmd;
 
-  if (open_loop)
-  {
-    // motor power (scale 0-1000)
-    int32_t right_power = right_speed / wheel_circumference * 60.0 / max_rpm * 1000.0;
-    int32_t left_power = left_speed / wheel_circumference * 60.0 / max_rpm * 1000.0;
-#ifdef _CMDVEL_DEBUG
-ROS_DEBUG_STREAM("cmdvel power right: " << right_power << " left: " << left_power);
-#endif
-    right_cmd << "!G 1 " << right_power << "\r";
-    left_cmd << "!G 2 " << left_power << "\r";
+  if (debug_){
+  ROS_DEBUG_STREAM("rpm right: " << rpm_msg.rpm_right << " left: " << rpm_msg.rpm_left);
+  ROS_DEBUG_STREAM("torque right" << rpm_msg.torque_right << " left " << rpm_msg.torque_left);
   }
-  else
-  {
-    // motor speed (rpm)
-    // The extra factor of 10.0 is a hack that is necessary for this to work in practice.
-    // Not really clear why we need to do this, but without it the commanded rpm
-    // is too low to make the motor move
-    int32_t right_rpm = right_speed / wheel_circumference * 60.0 * roboteq_rpm_scale;
-    int32_t left_rpm = left_speed / wheel_circumference * 60.0 * roboteq_rpm_scale;
-
-    wheel_rates_desired_msg.omega_left = left_rpm/60.0 * 2*M_PI;
-    wheel_rates_desired_msg.omega_right = right_rpm/60.0 * 2*M_PI;
-    wheel_rates_desired_msg.header.stamp = ros::Time::now();
 
-    wheel_rate_desired_pub.publish(wheel_rates_desired_msg);
-
-#ifdef _CMDVEL_DEBUG
-ROS_DEBUG_STREAM("cmdvel rpm right: " << right_rpm << " left: " << left_rpm);
-#endif
-    right_cmd << "!S 1 " << right_rpm << "\r";
-    left_cmd << "!S 2 " << left_rpm << "\r";
+  switch (drive_mode) {
+    case 0:
+      right_cmd << "!S 1 " << rpm_msg.rpm_right << "\r"; //in drive mode, we use the S command which is speed in RPM
+      left_cmd << "!S 2 " << rpm_msg.rpm_left << "\r";
+    case 5: //torque mode
+      right_cmd << "!TC 1 " << 100*rpm_msg.torque_right << "\r"; //in torque mode, we use the TC command which is torque in Nm*100
+      left_cmd << "!TC 2 " << 100*rpm_msg.torque_left << "\r";
+      break;
   }
-
-
-#ifndef _CMDVEL_FORCE_RUN
   controller.write(right_cmd.str());
   controller.write(left_cmd.str());
   controller.flush();
-#endif
 }
 
-void MainNode::cmdvel_setup()
+void MainNode::rpm_setup()
 {
-
   // stop motors
-  controller.write("!G 1 0\r");
+  controller.write("!G 1 0\r"); 
   controller.write("!G 2 0\r");
-  controller.write("!S 1 0\r");
-  controller.write("!S 2 0\r");
+  controller.flush();
```

Comment:
Looks like we now flush twice? Let's not do that if it's not necessary

Reply:
Removed